java-frontend/src/main/java/org/sonar/java/SonarComponents.java

```diff
@@ -325,4 +336,22 @@ public InputComponent project() {
     // TODO to be changed to context.project() once LTS 7.x has been released
     return context.module();
   }
+
+  public void collectUndefinedTypes(Set<String> undefinedTypes) {
+    this.undefinedTypes.addAll(undefinedTypes);
+  }
+
+  public void logUndefinedTypes() {
+    if (!undefinedTypes.isEmpty()) {
+      javaClasspath.logSuspiciousEmptyLibraries();
+      javaTestClasspath.logSuspiciousEmptyLibraries();
+
+      LOG.debug(String.format("Unresolved imports/types have been detected during analysis. Logging the %d first:", LOGGED_UNDEFINED_TYPES));
+      LOG.debug(undefinedTypes
+        .stream()
+        .sorted()
```

Comment:
As far as I understood `.sorted()` will sort strings in alphabetical order and `limit()` will show only first N.
In current implementation you will get first N elements after sorting them, so original order will not make sense and you probably don't need LinkedHashSet. 
If you want to get first N elements in the appearing order and then sort them, you need to swap calls  `limit` and `sorted`